### PR TITLE
security,sql,testutils,server,cli: Add testcases for client cert authentication.

### PR DESCRIPTION
**Pull request by BabuSrithar**
Data Driven tests covering:
* Certs without CN
* Certs with CN but also SAN
* Above certs in combination with default identity map (hba.conf) and custom
  identity map.

fixes None
Epic None

Release note: None

### CODE REVIEW

**Review by bdarnell — COMMENTED** [review]
> Reviewed 9 of 9 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/121004)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @BabuSrithar)

___
*[`pkg/security/securitytest/test_certs/regenerate.sh` line 22 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/121004#-NtuBIOhEZhlUJOnZiEf:-NtuBIOhEZhlUJOnZiEg:bfy249p) ([raw file](https://github.com/cockroachdb/cockroach/blob/854a65d616b55a567374ac69b0ee0006d1491cce/pkg/security/securitytest/test_certs/regenerate.sh#L22)):*
> ```Shell
> done
> 
> make generate PKG=./pkg/security/securitytest
> ```

The `make generate` line needs to go last so that everything gets embedded in the test binaries. 
___
*[`pkg/security/securitytest/test_certs/regenerate.sh` line 31 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/121004#-NtuBWiU7lS4jd0YAHq4:-NtuBWiU7lS4jd0YAHq5:b-pdgtl8) ([raw file](https://github.com/cockroachdb/cockroach/blob/854a65d616b55a567374ac69b0ee0006d1491cce/pkg/security/securitytest/test_certs/regenerate.sh#L31)):*
> ```Shell
> # san dns: only client certs
> openssl genrsa -out client.testuser_san_only.key
> openssl req -new  -key client.testuser_san_only.key -out client.testuser_san_only.csr -batch -subj /O=Cockroach -addext "subjectAltName = DNS:testuser"
> ```

Consider removing `O=Cockroach` from these. In this case we're interested in the CN vs SAN but we'll also have separate test cases for other fields in the DN.

**Review by BabuSrithar — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/121004)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @bdarnell)

___
*[`pkg/security/securitytest/test_certs/regenerate.sh` line 22 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/121004#-NtuBIOhEZhlUJOnZiEf:-NtuW_sUD49leG1qaHeQ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/854a65d616b55a567374ac69b0ee0006d1491cce/pkg/security/securitytest/test_certs/regenerate.sh#L22)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

The `make generate` line needs to go last so that everything gets embedded in the test binaries.
</blockquote></details>

Done.
___
*[`pkg/security/securitytest/test_certs/regenerate.sh` line 31 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/121004#-NtuBWiU7lS4jd0YAHq4:-NtuWQaH5zx6U3YZZod0:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/854a65d616b55a567374ac69b0ee0006d1491cce/pkg/security/securitytest/test_certs/regenerate.sh#L31)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

Consider removing `O=Cockroach` from these. In this case we're interested in the CN vs SAN but we'll also have separate test cases for other fields in the DN.
</blockquote></details>

Done.

**Review by bdarnell — COMMENTED** [review]
> Reviewed 3 of 14 files at r2, 44 of 46 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/121004)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @BabuSrithar, @DarrylWong, @rytaft, and @srosenberg)

___
*[`-- commits` line 2 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/121004#-Nvra6N8ASxPGRWYhTBt):*
We don't need to include every package with changed files in the first line of the commit message. I'd just do `security:`. 
___
*[`pkg/security/securitytest/test_certs/regenerate.sh` line 22 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/121004#-NvrQtjPBhAykq4vVB76:-NvrQtjPBhAykq4vVB77:b-5uyonh) ([raw file](https://github.com/cockroachdb/cockroach/blob/832574e994552a5ad9a4048ee7115344e97ae8a0/pkg/security/securitytest/test_certs/regenerate.sh#L22)):*
> ```Shell
> done
> 
> # Below section is all OpenSSL commands to generate the certs that require more options than
> ```

Two general comments that don't necessarily have to happen in this PR but could improve maintenance of these tests going forward:

1. It would be good if we didn't delete and regenerate all the files, but could operate on a subset when appropriate.
2. It's nice that `openssl ca` includes the textual decoding of the cert in its output. We could do the same in `cockroach cert` or run the certs through `openssl x509 -text` in this script. This makes it easier to tell what's changed in the binary blob. 
___
*[`pkg/sql/pgwire/auth.go` line 137 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/121004#-NvrRXDK2ywdTzFvdwIe:-NvrRXDK2ywdTzFvdwIf:by9bn3z) ([raw file](https://github.com/cockroachdb/cockroach/blob/832574e994552a5ad9a4048ee7115344e97ae8a0/pkg/sql/pgwire/auth.go#L137)):*
> ```Go
> 	c.sessionArgs.SystemIdentity = systemIdentity
> 
> 	log.Errorf(ctx, "dbUser: %v, systemIdentity: %v", c.sessionArgs.User, systemIdentity)
> ```

Remove this line
___
*[`pkg/sql/pgwire/testdata/auth/empty_hba` line 90 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/121004#-NvrXLNh99vhbUPVdKfA:-NvrXLNh99vhbUPVdKfB:b66kpbm) ([raw file](https://github.com/cockroachdb/cockroach/blob/832574e994552a5ad9a4048ee7115344e97ae8a0/pkg/sql/pgwire/testdata/auth/empty_hba#L90)):*
> ```
> # User with client cert, but the cert contains dbuser only in CN:
> # CN value = dbuser name
> # connect user=testuser database=defaultdb cert_name=testuser_cn_only force_certs show_system_identity
> ```

Uncomment this, or add a TODO comment with more details.

**Review by souravcrl — COMMENTED** [review]
PTAL. I have added a new failure test for SAN only certs.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/121004#-:-NvvmL_a0P-xetDcTtns:b-qqcg5e)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @bdarnell, @DarrylWong, @rytaft, and @srosenberg)

___
*[`-- commits` line 2 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/121004#-Nvra6N8ASxPGRWYhTBt):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

We don't need to include every package with changed files in the first line of the commit message. I'd just do `security:`.
</blockquote></details>

updated to only security package
___
*[`pkg/security/securitytest/test_certs/regenerate.sh` line 22 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/121004#-NvrQtjPBhAykq4vVB76:-NvvOoCi9AzECZ0LoXFv:b6tl6a4) ([raw file](https://github.com/cockroachdb/cockroach/blob/832574e994552a5ad9a4048ee7115344e97ae8a0/pkg/security/securitytest/test_certs/regenerate.sh#L22)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

Two general comments that don't necessarily have to happen in this PR but could improve maintenance of these tests going forward:

1. It would be good if we didn't delete and regenerate all the files, but could operate on a subset when appropriate.
2. It's nice that `openssl ca` includes the textual decoding of the cert in its output. We could do the same in `cockroach cert` or run the certs through `openssl x509 -text` in this script. This makes it easier to tell what's changed in the binary blob.
</blockquote></details>

created a ticket for this: https://cockroachlabs.atlassian.net/browse/CRDB-38047
___
*[`pkg/sql/pgwire/testdata/auth/empty_hba` line 90 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/121004#-NvrXLNh99vhbUPVdKfA:-NvvLs9Y2HWAGxz4F01Z:ba2blbz) ([raw file](https://github.com/cockroachdb/cockroach/blob/832574e994552a5ad9a4048ee7115344e97ae8a0/pkg/sql/pgwire/testdata/auth/empty_hba#L90)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

Uncomment this, or add a TODO comment with more details.
</blockquote></details>

uncommented post rebasing and removing validation fn `validateCockroachCertificate` as it is only a sanity test and with the remapping features this check is no longer required 
___
*[`pkg/sql/pgwire/auth.go` line 137 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/121004#-NvrRXDK2ywdTzFvdwIe:-NvvLkQlBfZoJp3xQ81M:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/832574e994552a5ad9a4048ee7115344e97ae8a0/pkg/sql/pgwire/auth.go#L137)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

Remove this line
</blockquote></details>

done

File: pkg/security/securitytest/test_certs/ca-client-tenant.crt [hunk 1]
```diff
@@ -1,19 +1,19 @@
...
-tKbkgu0Min7UglUEP2pOtaw4JxKm2bpci8YlIxiP1v7UNcdNt4VkQHlwafV7hfXw
-dvwQgk4qcdATddnF16v3btCpFlBK0cxb5GUCAwEAAaNFMEMwDgYDVR0PAQH/BAQD
-AgLkMBIGA1UdEwEB/wQIMAYBAf8CAQEwHQYDVR0OBBYEFAWfIwqohtRUNFlkDqkX
-ML9ratkrMA0GCSqGSIb3DQEBCwUAA4IBAQCvvIJc2Pgzfqa+0EsjAZfDBr4t5G2Y
-RSEviB3qycvDvvUnEIA+9q5eB1pvwbq3F62GgqY3q0viyJmWayVCEIRSZvbyPbQq
-7z85KgYofYYgUm3xY9UVDp79NvBatJAN96pX1q14IcsXQTk3pKx5csuRiSxy2hQa
-j/l1BVvWwSv8YxTpHC4vBbRtEuDiT837DvgQQJxuqt0e9xhYBASYu9XkhpxJdIgT
-9pikiaEVfXDHHRxapw/3znaH6WABYEgF0Jgk/yFe+MxPB9UVSt8m0QNfiL2uVsaQ
-fRCS8f4C2s5uDL55LlEsuRi4mN0M7XUAPPiWo5UdxFe6af/mGvRvHxpT
+MIIDJTCCAg2gAwIBAgIQHydtwmtLPitWCtI/kEJvQTANBgkqhkiG9w0BAQsFADAr
```
Comment (BabuSrithar) [inline pkg/security/securitytest/test_certs/ca-client-tenant.crt:2]: Is this change required or was it just a change because we ran regenerate.sh ?

**Review by souravcrl — COMMENTED** [review]
all required CI are passing post updating unit tests which depended on error cases for the `validateCockroachCertificate` function.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/121004#-:-Nw4m_ZG0cWfZIwNB2K2:bkcrtkh)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @BabuSrithar, @bdarnell, @DarrylWong, @rytaft, and @srosenberg)

___
*[`pkg/security/securitytest/test_certs/ca-client-tenant.crt` line 2 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/121004#gh-1574571689:-Nw4mJWxCnpF8fVBcclE:b1spxp4) ([raw file](https://github.com/cockroachdb/cockroach/blob/f0d6cbac11232e719d1a15c42e8d248f23c97bc6/pkg/security/securitytest/test_certs/ca-client-tenant.crt#L2)):*
<details><summary><i>Previously, BabuSrithar (BabuSrithar) wrote…</i></summary><blockquote>

Is this change required or was it just a change because we ran regenerate.sh ?
</blockquote></details>

This got regenerated on running `regenerate.sh`. There were a few comments regarding the script which we need to take up and I have a created a ticket to track this: [https://cockroachlabs.atlassian.net/browse/CRDB-38047](https://cockroachlabs.atlassian.net/browse/CRDB-38047)

**Review by bdarnell — APPROVED** [review]
> Reviewed 2 of 37 files at r4, 1 of 36 files at r6, 4 of 39 files at r7, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/121004)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @BabuSrithar, @DarrylWong, @rytaft, and @srosenberg)

**Comment by souravcrl** [comment]
TFTR!

bors r+
